test/externalTests/foundry.py

```diff
@@ -0,0 +1,178 @@
+#!/usr/bin/env python3
+
+# ------------------------------------------------------------------------------
+# This file is part of solidity.
+#
+# solidity is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# solidity is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with solidity.  If not, see <http://www.gnu.org/licenses/>
+#
+# (c) 2023 solidity contributors.
+# ------------------------------------------------------------------------------
+
+import os
+import re
+from pathlib import Path
+from shutil import which
+from string import Template
+from typing import Tuple
+import requests
+
+from external_test import AVAILABLE_PRESETS
+from external_test import settings_from_preset, run_cmd
+from external_test import TestConfig, TestRunner, ExternalTest
+
+# TODO: Add hardhat runner
+# TODO: Add truffle runner
+
+class FoundryRunner(TestRunner):
+    """Configure and run Foundry-based projects"""
+
+    profile_tmpl = Template("""
+[profile.${name}]
+gas_reports = [\"*\"]
+auto_detect_solc = false
+solc = \"${solc}\"
+evm_version = \"${evm_version}\"
+optimizer = ${optimizer}
+via_ir = ${via_ir}
+[profile.${name}.optimizer_details]
+yul = ${yul}
+""")
+
+    foundryup_url = "https://raw.githubusercontent.com/foundry-rs/foundry/master/foundryup/foundryup"
+    foundry_repo = "https://github.com/foundry-rs/foundry.git"
+
+    def __init__(self, config: TestConfig, setup_fn=None, compile_fn=None, test_fn=None):
+        self.config = config
+        self.setup_fn = setup_fn
+        self.compile_fn = compile_fn
+        self.test_fn = test_fn
+        self.env = os.environ.copy()
+        # Note: the test_dir will be set on setup_environment
+        self.test_dir: Path = None
+
+    def setup_environment(self, test_dir: Path, from_src: bool = False):
+        """Configure the project build environment"""
+
+        self.test_dir = test_dir
+        os.chdir(test_dir)
+
+        print("Configuring Foundry building environment...")
+        if which("forge") is None:
+            if not which("cargo"):
+                raise RuntimeError("Cargo not found.")
+            foundry_dir = test_dir / "foundry"
+            foundry_bin_dir = foundry_dir / "bin"
+            foundry_bin_dir.mkdir(parents=True, exist_ok=True)
+            self.env.update({"FOUNDRY_DIR": foundry_dir.resolve()})
+            if from_src:
+                ExternalTest.download_project(foundry_dir, self.foundry_repo, "branch", "master")
```

Comment:
Why do you need to download foundry repo if you already installed it with `foundryup`?

Reply:
I initially thought in provide both options, since we could decide to use some specific branch of it. But I will just move the installation process to the CI and use the foundryup script with the selected foundry release

Reply:
Yeah, we likely won't need that. Even with Hardhat we managed to get by sticking only to official releases. With Foundry it should be even less of a problem. In the npm ecosystem it's too easy for people to treat their tools as dependencies, keep them frozen and stay on ancient versions forever. I haven't seen them hard-coding a specific Foundry version on the other hand.